### PR TITLE
feat: login functionality

### LINKED ISSUE #2: User management
As a dev
I want to user management tools
So can manage my users

Use Pocket base?

### PR DESCRIPTION
**Pull request by namatoj**
Added functionality to use the JWT in a session cookie as mentioned [here](https://github.com/namatoj/sociallink/issues/2)

Added the following routes:
- /
- /login/
- /logout/

If sending a POST request to `/login/` containing the form-data values `email` and `password` while it is served using https corresponding to a user in the user collection the cookie `token` will be set.

@simondmansson I'm a bit unsure about how we want to structure the routes and the middleware function. It quite fast became rather messy. :-) I'm open for suggestions. Othewise I will leave this PR for a bit and come back when I have more inspiration for structure.

### CODE REVIEW

**Comment by simondmansson** [comment]
I think we should start flat and add more folders and re-group as we go and learn more about what we are building.
We could set it up like this

#### pkg/web/app.go
creates the pocketbase app and responsible for grouping and registering handlers register handlers, e.g
```go
func App() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.Use(auth.LoadAuthContextFromCookie(app))
		e.Router.GET("/*", apis.StaticDirectoryHandler(os.DirFS("./pb_public"), false))
		e.Router.POST("/login/", auth.LoginHandler)
		e.Router.GET("/logout/", auth.LogoutHandler)
		e.Router.GET("/", start.viewHandler)
	})

	return app
}
```
#### pkg/web/auth.go 
have handler funcs for login, logout, cookie management e.g 
```go
func LogoutHandler(c echo.Context) error {
	c.SetCookie(&http.Cookie{
		Name:     "token",
		Value:    "",
		Path:     "/",
		MaxAge:   -1,
		Secure:   true,
		HttpOnly: true,
	})

	return c.HTML(http.StatusOK, "logged out")
}
```
#### cmd/web/main.go 
imports the app and starts it. Might do some additional flag / config parsing in the future.

**Comment by simondmansson** [comment]
This video gives some great tips on code structure https://www.youtube.com/watch?v=oL6JBUk6tj0

**Comment by simondmansson** [comment]
Also by breaking out the handlers we can setup auth_test.go and write some simple  "given x is in the context the handler should do y" tests.

**Comment by namatoj** [comment]
> This video gives some great tips on code structure https://www.youtube.com/watch?v=oL6JBUk6tj0

That was inspiring! Got me excited about the hexagonal architecture.

> Also by breaking out the handlers we can setup auth_test.go and write some simple "given x is in the context the handler should do y" tests.

Agree! I'll break out the handlers to somewhere in `pkg`. To be continued!

**Comment by namatoj** [comment]
Something a bit like this maybe?

I had some troubles with some of the details in your suggestion. `auth.LoginHandler` suggest that `auth` is in another package (right?). But the directory structure that `auth.go` should be in the same directory as `app.go` which didn't vibe with the `package` keyword on the top row.

This lead me to take the easy way of putting `auth.go` and `app.go` in the `web` package. Let me know if I misunderstood anything or if you have more feedback on how to improve this.

File: pkg/web/auth.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	return c.HTML(http.StatusOK, "logged out")
+}
+
+func LoadAuthContextFromCookie(app core.App) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			tokenCookie, err := c.Request().Cookie("token")
+			if err != nil || tokenCookie.Value == "" {
+				return next(c) // no token cookie
+			}
```
Comment (simondmansson) [inline pkg/web/auth.go]: We can set isGuest here in the context and then access the value in other handlers as needed.
The context key can be declared after imports as 
```go
const (
	ContextAuthIsGuestKey string = "isGuest"
)
```
```suggestion
			if err != nil || tokenCookie.Value == "" {
			        c.Set(ContextAuthIsGuestKey, true)
				return next(c) // no token cookie
			}
```

File: pkg/web/app.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+	return app
+}
+
+func RootHandler(c echo.Context) error {
+	info := apis.RequestInfo(c)
+	admin := info.Admin       // nil if not authenticated as admin
+	record := info.AuthRecord // nil if not authenticated as regular auth record
+
+	isGuest := admin == nil && record == nil
```
Comment (simondmansson) [inline pkg/web/app.go]: If we set isGuest in the context we can fetch it here and check if the value is true or false. `v`can be renamed to something more descriptive 
```suggestion
      isGuest := c.Get(ContextAuthIsGuestKey)
	v := cast.ToBool(isGuest)
	if v == true {
		return c.HTML(http.StatusOK, "not logged in")
	}
```

**Review by simondmansson — COMMENTED** [review]
Looks good. I have two comments about isGuest. I'm not sure how we are going to use it in the future but always having the calculated value in the context will be handy i think.

File: pkg/web/app.go (hunk 2 above)
Comment (simondmansson) [reply pkg/web/app.go]: The C.get then cast into bool could be a helper func in the future if needed.

Comment (namatoj) [reply pkg/web/app.go]: Thank you for the feedback! I've pushed new changes now.

**Review by simondmansson — APPROVED** [review]

**Pull request merged**